Add custom configurations for enterprise

kanishk98 opened the pull request:
---
<!--
Remove the fields that are not appropriate
Please include:
-->

**What's this PR do?**

Considers a new custom config file for managing user-level settings of the app. We expect that the admin deploying Zulip at enterprise will `echo` the filepath for this config file into `settings.json` and then we can read this as replacements for settings in the app. 

**Any background context you want to provide?**

A major request by enterprise users is that they should be able to regulate user-level settings by maintaining a `custom_settings.json` file at a system-level settings. Following the discussion [here](https://chat.zulip.org/#narrow/stream/16-desktop/topic/high.20availability/near/701040), I have implemented a simple method to read enterprise settings using `fs`.
This assumes that the admin will `echo` the filepath of the custom config file as per their choice in the script they write and the app will just read that.

I also realised that some settings (say auto updates) should be disabled by admins but some settings (like showing the sidebar) should be accessible to the user. So I've added an `admin` field to every config item. By default, it is false, which means that admin access to change that field is not required. Handled that in ConfigUtil's `setConfigItem` directly.

**Slated improvements**

I've sent in this draft PR to make sure everyone invested in this functionality (quite a few people) can provide feedback as we go along. That being said, I plan to add the following features over the next couple of days:

- [x] 1. Extend the switches' permissions regulation to other settings fields and dialog boxes
- [x] 2. Extend similar permissions to preset organizations
- [x] 3. Investigate options we have in terms of granularity and add support for the relevant ones
- [x] 4. Document the new files made and the process of editing custom settings

**You have tested this PR on:**
  - [x] Windows
  - [ ] Linux/Ubuntu
  - [ ] macOS

akashnimare commented on app/renderer/js/utils/custom-config-util.js:
```diff
@@ -0,0 +1,93 @@
+		return (value !== undefined);
+	}
+
+	removeConfigItem(key) {
```
@kanishk98 I don't think we will be able to use this function since we won't have rights to remove/delete the data from that particular directory. cc, @priyank-p.

kanishk98 replied:
Okay, makes sense. I'll remove this method.

priyank-p replied:
Also, I don't think we will need the `node-json-db` here. We just need to attempt to read the json which you can do with with reading the file with `fs` and `JSON.parse`.

priyank-p commented on app/renderer/js/main.js:
```diff
@@ -147,6 +149,20 @@ class ServerManagerView {
 	}
 
+	initEnterpriseSettings() {
+		const filepath = ConfigUtil.getConfigItem('enterpriseFilepath', null);
```
The whole chunk here is wrong; we want something like
```javascript
const possibleFilePaths = [
  '/etc/zulip-desktop/enterprise.json', // mac and linux
  'C://Program Files/Zulip-Desktop-Config' // windows
].map(filepath => path.resolve(filepath);
```
and then check each file path by looping through the each of them. Also import the node.js path module here to make sure it parse the file path so they are cross-platform safe.

priyank-p commented on app/renderer/js/main.js:
```diff
@@ -147,6 +149,20 @@ class ServerManagerView {
+		if (filepath) {
+			fs.readFile(filepath, (err, data) => {
+				if (err) {
+					logger.reportSentry(err);
```
This is wrong. If the config file doesn't exist it will report that to sentry and spam it with these reports. It is ok for the config file to not exist.

priyank-p commented on app/renderer/js/main.js:
```diff
@@ -147,6 +149,20 @@ class ServerManagerView {
+					return;
+				}
+				const enterpriseSettings = JSON.parse(data);
+				console.log(enterpriseSettings);
```
Remove this debug `console.log`

priyank-p commented on app/renderer/js/utils/config-util.js:
```diff
@@ -50,7 +50,7 @@ class ConfigUtil {
 			return defaultValue;
 		} else {
-			return value;
+			return value.value;
```
What's the story behind this change? Can you add a comment here or in the commit message about this.

priyank-p commented on app/renderer/js/utils/config-util.js:
```diff
@@ -66,7 +66,24 @@ class ConfigUtil {
+			// if admin is set to true, user is not allowed to change that setting
+			return;
+		}
+		if (this.isConfigItemExists(key) && (value === null || value.admin === undefined)) {
```
All of these conditionals look fragile can you explain what these are if statement are doing for all the cases. I definitely feel there is cleaner and easier way to write these here. Also, I recommend checking `value.admin !== true` here.

priyank-p commented on app/renderer/js/utils/enterprise-util.js:
```diff
@@ -0,0 +1,53 @@
+const fs = require('fs');
```
This change here should have been done in the first commit; I ended up reviewing a change that you end up removing in the end.

priyank-p commented on app/renderer/js/utils/enterprise-util.js:
```diff
@@ -0,0 +1,53 @@
+		} catch (err) {
+			logger.error('Error while JSON parsing enterprise_config.json: ');
+			logger.error(err);
+			logger.reportSentry(err);
```
See the comment I added earlier for reporting this.

priyank-p commented on app/renderer/js/utils/enterprise-util.js:
```diff
@@ -0,0 +1,53 @@
+	reloadDB() {
+		let enterpriseJsonPath = null;
+		if (process.platform === 'win32') {
+			enterpriseJsonPath = 'C:\\Program Files\\Zulip\\enterprise_config.json';
```
I would avoid using `Zulip` directory since if the admin installs a desktop app it ends up here. Also see the comment I made earlier (possible would be marked as outdated) about using path module and the file paths.

akashnimare replied:
Would be great if we can use the electron's userData api here to get the correct path of the installed app.

priyank-p commented on app/renderer/js/utils/config-util.js:
```diff
@@ -65,26 +66,11 @@ class ConfigUtil {
-		if (this.isConfigItemExists(key) && this.getConfigItem(key).admin) {
-			// if admin is set to true, user is not allowed to change that setting
-			return;
+	setConfigItem(key, value, override = false) {
```
What the point of having the optional override parameter, for future use?

kanishk98 replied:
While initializing default settings for the first time, the app may read a few admin-only fields from the custom config file that will need to reflect in the user's `settings.json`. To keep code simple in `setConfigItem()`, I just set the optional override parameter to true when initializing the app. Requires less/no changes elsewhere in the app that this function is used.

kanishk98 replied:
Here is the code in `initDefaultSettings()`:
```javascript
if (EnterpriseUtil.isConfigItemExists(i)) {
	ConfigUtil.setConfigItem(i, EnterpriseUtil.getConfigItem(i), true);
} else {
	ConfigUtil.setConfigItem(i, settingOptions[i]);
}
```

priyank-p replied:
OK, make sense.

kanishk98 replied:
Thanks for the review. I'll clean up git history and make the changes required.

priyank-p reviewed (commented):
@kanishk98 sorry for delayed review! I went through half of the changes didn't get to review the preset orgs commit and later.

First, I recommend you clean up you git history, you made a lot of changes in first couple of commit that you changed in later commit. I just waste of time to review those changes.

kanishk98 commented:
@akashnimare @priyank-p it'll be great if either of you can find some time to review the recent changes. 

I'll be making the required UI changes that should correspond to disallowed, admin-only actions and setup stuff (`initPresetOrgs()`) in a couple of hours.

priyank-p commented on app/renderer/js/main.js:
```diff
@@ -152,6 +158,34 @@ class ServerManagerView {
+		const presetOrgs = EnterpriseUtil.getConfigItem('presetOrganizations', []);
+		const addDomainPromises = [];
+		for (const url in presetOrgs) {
+			if (!DomainUtil.duplicateDomain(presetOrgs[url])) {
```
Let's do:
```javascript
if (DomainUtil.duplicateDomain(presetOrgs[url])) {
  continue;
}
```

here and the rest the stuff at bottom since the chunk inside the `if` statement is huge.

priyank-p commented on app/renderer/js/main.js:
```diff
@@ -152,6 +158,34 @@ class ServerManagerView {
 		this.toggleSidebar(showSidebar);
 	}
 
+	initPresetOrgs() {
```
Let make this a `async` function so we can use `await` and avoid the complex `then` chaining!

kanishk98 replied:
Ah okay. I'd not seen `async` used anywhere so far, I just assumed we didn't like it in the codebase for some reason. I'll change it.

priyank-p replied:
It not visible in the codebase because when most of the stuff was written `async` wasn't supported. Try using whenever possible it will simplify the code significantly. The `then` changing gets hard to follow.

priyank-p commented on app/renderer/js/utils/enterprise-util.js:
```diff
@@ -0,0 +1,69 @@
+		}
+	}
+
+	isConfigItemExists(key) {
```
For the function name here `isConfigItemExists` sounds odd; let's make it `configItemExists`.

priyank-p commented on app/renderer/js/main.js:
```diff
@@ -142,7 +144,11 @@ class ServerManagerView {
 		for (const i in settingOptions) {
 			if (ConfigUtil.getConfigItem(i) === null) {
-				ConfigUtil.setConfigItem(i, settingOptions[i]);
+				if (EnterpriseUtil.isConfigItemExists(i)) {
```
The `i` here is a setting name here right? I think it makes sense to change the `i` variable `setting` in a separate commit.

priyank-p replied:
Add a comment here stating what this does. I think this initializes default based on enterprise config. If so mention that in a comment.

priyank-p commented on app/renderer/js/utils/enterprise-util.js:
```diff
@@ -0,0 +1,68 @@
+	}
+}
+
+module.exports = new EnterpriseUtil();
```
Missing ending newline here.

priyank-p commented on app/renderer/js/utils/enterprise-util.js:
```diff
@@ -0,0 +1,68 @@
+
+	getConfigItem(key, defaultValue = null) {
+		this.reloadDB();
+		const value = this.enterpriseSettings[key];
```
```javascript
const { value } = this.enterpriseSettings[key];
```

here, so we don't need to do `value.value` below it would be just `value`.

priyank-p commented on app/renderer/js/utils/enterprise-util.js:
```diff
@@ -0,0 +1,68 @@
+			return defaultValue;
+		} else {
+			return value.value;
+		}
```
This `if else` block could be just `return (value || defaultValue)`.

priyank-p commented on app/renderer/js/utils/enterprise-util.js:
```diff
@@ -25,17 +26,17 @@ class EnterpriseUtil {
+			enterpriseJsonPath = 'C:\\Program Files\\Zulip-Desktop-Config\\enterprise_config.json';
 		} else {
-			enterpriseJsonPath = '/etc/zulip/enterprise_config.json';
+			enterpriseJsonPath = '/etc/zulip-desktop-config/enterprise_config.json';
```
These changes should have been part of the previous change.

@kanishk98 [these docs](https://zulip.readthedocs.io/en/latest/git/fixing-commits.html#fixing-older-commits) should help you out with this stuff if you are having trouble or just not confident in doing git rebase.

kanishk98 replied:
Sure sure. I forgot to rebase these changes.

priyank-p commented on app/renderer/js/utils/enterprise-util.js:
```diff
@@ -25,17 +26,17 @@ class EnterpriseUtil {
-			logger.error('Error while JSON parsing enterprise_config.json: ');
-			logger.error(err);
-			logger.reportSentry(err);
+			logger.log('Error while JSON parsing enterprise_config.json: ');
```
ditto :)

priyank-p commented on app/renderer/js/utils/domain-util.js:
```diff
@@ -103,11 +112,14 @@ class DomainUtil {
 			return Promise.reject('This server has been added.');
 		}
+		if (EnterpriseUtil.isAdminOnly('presetOrganizations') && override === false) {
+			return Promise.reject('Adding organizations is a restricted operation. Please contact your system administrator.');
```
I don't think we want to restrict users from adding a new organization. The most we should restrict them to do is removing a preset org (and even that I feel like should be a new setting).

cc @akashnimare

kanishk98 replied:
Okay, I'll wait for more discussion on this.

akashnimare replied:
Agreed.

priyank-p reviewed (commented):
@kanishk98 Left a lot of cleanup comment for code, some are simple whereas others are more involved changes.

Also left a comment about the preset org commit. Can you briefly explain what everything does? For example, adding what config affect what parts of the app, etc.

kanishk98 commented:
> Left a lot of cleanup comment for code, some are simple whereas others are more involved changes.

Sure, I'll clean this up. 

> Can you briefly explain what everything does? For example, adding what config affect what parts of the app, etc.

I'll be documenting this, but here's the sample .json file I've been using to test this PR:
```json
{
	"autoHideMenubar": {
		"value": false,
		"isAdminOnly": true
	}, 
	"useSystemProxy": {
		"value": false, 
		"isAdminOnly": false
	},
	"presetOrganizations": {
		"value": ["https://kanishk.com"],
		"isAdminOnly": true
	}
}
```

We have support for the settings handled by `ConfigUtil` and `presetOrganizations` so far. If a setting's `isAdminOnly` field is set, the user can't change that setting, and if it's the `presetOrganizations` field that's admin-only, then the user can neither add nor delete organizations. (Will be changing this based on discussion above).

kanishk98 commented:
Fixed up relevant commits directly and force-pushed. Continuing further improvements now. :)

kanishk98 commented:
I realized I hadn't handled the case when an admin might want to change the preset organizations they had decided on initially. Now when that happens, I think the app should silently add the new organizations in the background and if the addition is successful, then it should prompt the user to reload the app (similar to how we notify the user that an update has been downloaded). 

If an organization has been deleted and the app loads after this config change, then I think we should immediately reload the app without waiting for the user to confirm. 

I think I've cleaned up my commits as required by @priyank-p, so I'll go ahead and start writing these changes. Please let me know what you think.

kanishk98 commented:
I've acted on the comment above and also fixed something we overlooked wrt admin-only settings earlier: the user would've been able to go into `settings.json` and change stuff there. So I altered [this commit](https://github.com/zulip/zulip-desktop/pull/681/commits/577b282ad9b510774fb38ef829e4400ea16d2f90) and rebased to accomodate this change. Here's the new line of code when considering default settings:

```javascript
for (const setting in settingOptions) {
			if (ConfigUtil.getConfigItem(setting) === null || EnterpriseUtil.isAdminOnly(setting)) {
```

This forces the app to reset any admin-only settings each time it is loaded.

kanishk98 commented:
I have a few UI-specific concerns:

1. When adding preset organizations, the app's UI doesn't indicate anything. This could be solved via 2 ways, I think - one is by showing a rotating icon overlaying the app or (my preference) a rotating icon in the sidebar similar to the one implemented in #674. 
2. When trying to change an admin-only setting, we could either point out to the user every time that the setting is fixed (not a good idea), show the switch snapping back to its config-oriented setting by using a timeout, or we could shake the app angrily (I believe macOS apps do something similar when they're experiencing an error, not sure). 
3. What kind of UI should we present to the admins for modifying these settings, if any? I believe we discussed a web UI earlier on the czo private group, but I'd like to be a little more specific. 

@rishig please let us know what you think.

priyank-p commented:
@kanishk98 I started a thread on [#desktop Enterprise Configuration UI](https://chat.zulip.org/#narrow/stream/16-desktop/topic/Enterprise.20Configuration.20UI) for this.

priyank-p commented:
I pushed a version that:
- Add a commit on top that just adds `enterprise-util.js`.
- Removes `require("process")` since its unnecessary and weird.
- In the `reloadDB` function you were reading the file and parsing it in the `try` block and in the `catch` block you are logging that there is error parsing json which could be wrong  since the `fs.readFile` can throw error if the file doesn't exist. I handled the case by moving the try block inside the if condition that check `fs.

kanishk98 commented:
A list of changes I've made since the last review:

1. Modified [ee937f](https://github.com/zulip/zulip-desktop/pull/681/commits/ee937f9250a31ea3e3dd586f6f182a091ee9a9af) to use `async-await` instead of `then` chains. 
2. Detect changes in the config file so that the app automatically removes any organisations not in the config if preset organisations are set as admin-only and also adds apps in the background if new orgs are added later.
3. Show `cursor: not-allowed` on admin-only fields.
4. Disable required configurable settings in the menu if setting is admin-only. 
5. Show the standard loading gif when the app has no domains added and orgs are being added in the background.

priyank-p commented on app/renderer/js/main.js:
```diff
@@ -159,6 +160,31 @@ class ServerManagerView {
+			}
+			try {
+				// eslint-disable-next-line no-await-in-loop
+				const serverConf = await DomainUtil.checkDomain(presetOrgs[url], {override: true});
```
Using `await` in for loop, as the linting rule enforces, is bad idea. Essentially you are waiting for one operation to finish while other could have been in progress. You should split the code inside of the try block into a async function, then on each loop call the function which will return a promise and store it in a array. Then after the loop do `await Promise.all(<array-of-promise>);`.

kanishk98 replied:
You're right, I should've taken care of that. 
I'll resolve with `false` when there's an error and with `true` when the domain gets added successfully, since we want to reload if at least one org gets added.

kanishk98 commented:
Let me know if any more changes are required.

kanishk98 committed:
docs: Add enterprise docs for preset orgs.

kanishk98 commented:
I've changed the PR to allow adding orgs at all times, and show a prompt disabling removal of the orgs specified by the admin, if they're set to be admin-only. 
I also added a doc explaining the enterprise config file. 

@priyank-p do you think we should consider using an npm module to check if the user is root, and then provide an auth prompt or override access depending on that? (Currently, we'll need the user to change the config file and then reload the app). 

@akashnimare I'd mentioned writing a script for generating the config file earlier - I think accepting input on the URLs required and the admin only setting is the best idea for that. (Similar to what we get when we run `$ npm init`)

kanishk98 commented:
> you think we should consider using an npm module to check if the user is root, and then provide an auth prompt or override access depending on that?

I think this is a good idea. I added [sudo-prompt](https://www.npmjs.com/package/sudo-prompt) and this is the flow for removing an org now if preset orgs are admin-only:

![sudo](https://user-images.githubusercontent.com/24617297/59220393-7d25a680-8be2-11e9-9f24-930659d635bf.gif)

One complication that this brings is that we'll need to remember the orgs which were deleted by admins so we don't keep removing them. 

(would just editing the config file directly from here be a good idea?)

rishig commented:
I think we should just show an error message saying that this organization is configured by /path/to/config/file, and say to contact their Zulip administrator.

kanishk98 commented:
> I think we should just show an error message saying that this organization is configured by /path/to/config/file, and say to contact their Zulip administrator.

Yeah, that's what Akash and I had considered earlier, but I figured it might be easier for the admins (and certainly the devs) using the app to just get access temporarily to the setting. 
Let me know if you think I should still revert this commit, though.

rishig commented:
No comment on the dev experience part, but I think the production version of this feature should not have the ability to change the config settings without directly changing the file, since adding a UI for it adds some implicit expectations that will be a big can of worms.

kanishk98 commented:
> since adding a UI for it adds some implicit expectations that will be a big can of worms.

Fair enough, I'll drop the commit. Thanks!

priyank-p commented on app/renderer/js/main.js:
```diff
@@ -159,6 +160,35 @@ class ServerManagerView {
+			return true;
+		} catch (err) {
+			logger.error(err);
+			logger.error('Could not add ' + domain + '. Please contact your system administrator.');
```
Hmm, do you mean to put this in dialog? Just checking since this is going to log file and this seems more like user-facing error message.

kanishk98 replied:
No, I figured that would be annoying, because any domains not added yet and present in the config file will be checked and added, so any incorrect domains will keep opening up a dialog every time the app loads. 
We could probably show a dialog the first time that this fails and then only log the messages from then on. What do you think?

kanishk98 replied:
@priyank-p I've added a commit that shows an error dialog when none of the organizations in `presetOrgs` can be added. If you think we need to be more detailed, then let's use this dialog to list all domains that failed to load. 
(If we just replace the `logger` statement with the `dialog`, then every failing org will trigger an error dialog).

priyank-p commented on app/resources/messages.js:
```diff
@@ -22,6 +22,13 @@ class Messages {
 		\nYou can click here if you'd like to proceed with the connection.`;
 	}
+
+	failEnterpriseOrgError(length) {
```
`enterpriseOrgError` along sound good to me as the function name here.

kanishk98 replied:
Alright, no problem. I'll change that.

kanishk98 committed:
enterpise: Add enterprise-util for custom configs.

priyank-p commented on app/renderer/css/preference.css:
```diff
@@ -453,6 +453,10 @@ i.open-tab-button {
 }
 
+input.toggle-disabled + label {
+    cursor: not-allowed;
```
A good followup for this commit might be showing a tooltip or text showing why it's not allowed.

priyank-p commented on app/renderer/js/main.js:
```diff
@@ -608,7 +608,7 @@ class ServerManagerView {
 						}, response => {
 							if (response === 0) {
-								if (EnterpriseUtil.isAdminOnly('presetOrganizations')) {
+								if (EnterpriseUtil.isAdminOnly('presetOrganizations') && this.presetOrgs.includes(DomainUtil.getDomain(index).url)) {
```
This commit probably needs to be squash with the commit that introduced the feature.

priyank-p commented on docs/Enterprise.md:
```diff
@@ -0,0 +1,26 @@
+It must be named `enterprise_config.json` in both cases. 
+
+To specify the preset organization you want to add for other users, you will need to
+add below to the `enterprise_config.json`, replace `https://chat.zulip.org` to the
```
```suggestion
add the json shown below to the `enterprise_config.json`, replace `https://chat.zulip.org` to the
```

priyank-p commented on app/renderer/js/main.js:
```diff
@@ -153,6 +162,76 @@ class ServerManagerView {
+			}
+		} else {
+			// no enterprise domain added
+			const { title, content } = Messages.failEnterpriseOrgError(this.presetOrgs.length);
```
```suggestion
			const { title, content } = Messages.enterpriseOrgError(this.presetOrgs.length);
```

This should be `enterpriseOrgError` since that was added to `Message` in this commit?

priyank-p commented on app/renderer/js/utils/enterprise-util.js:
```diff
@@ -0,0 +1,63 @@
+		instance = this;
+
+		this.reloadDB();
+		return instance;
```
Remove this return statement.
```suggestion
		
```

priyank-p reviewed (commented):
@kanishk98 this looks great but posted bunch of comments. Also not that you are missing `.` at the end of commit messages can you also fix that.

kanishk98 commented:
Included changes suggested by @priyank-p. Also removed `isAdminOnly()` requirements as suggested by Tim and Rishi on czo earlier. 

Here's a screenshot of the tooltip suggested: 

![Screenshot from 2019-06-29 05-57-42](https://user-images.githubusercontent.com/24617297/60377361-bf007a80-9a33-11e9-95fc-929942de2497.png)

rishig commented on app/renderer/js/pages/preference/base-section.js:
```diff
@@ -18,7 +18,7 @@ class BaseSection extends BaseComponent {
 
 	generateOptionTemplate(settingOption, disabled) {
-		const label = disabled ? `<label class="disallowed" />` : `<label/>`;
+		const label = disabled ? `<label class="disallowed" title="Setting locked by admin."/>` : `<label/>`;
```
@timabbott, thoughts on what this string should be? I like the "Setting locked by X" format, but not sure what to call the person doing the locking here.

timabbott replied:
I think "Setting locked by your system administrator." is probably correct, since "system administrator" pretty unambiguously refers to folks with root access on your computer.

rishig replied:
yeah, I like that.

kanishk98 commented:
> I think "Setting locked by your system administrator." is probably correct

Great, changed the message. :)

kanishk98 committed:
enterprise: Add support for preset orgs.

Enable configuration of preset organizations in the enterprise config
file and load them when app loads.
Partially valid array of URLs accepted while loading app.

kanishk98 committed:
enterprise: Detect changes in orgs config.

Allow app to identify if presetOrgs have changed in enterprise config
and change domains based on user response.

kanishk98 committed:
enterprise: Show loading gif view.

When no other orgs have been added and presetOrgs have been specified in
enterprise config, loading gif will appear.

kanishk98 committed:
enterprise: Show error when all preset orgs fail.

On app load, if all the preset orgs fail to get added to the server,
then we show an errorBox.

kanishk98 committed:
enterprise: Stop loading gif if no domains added.

If no domains have been added, stop the loading gif and open the new
server form.

kanishk98 committed:
enterprise: Raise error when removing orgs.

When a user tries to remove a preset org, we raise an error asking them
to contact their sys admin.

kanishk98 committed:
enterprise: Init enterpriseSettings from file.

Read enterprise settings while loading the app and give preference to
global config.

kanishk98 committed:
enterprise: Disable autoUpdate on admin setting.

kanishk98 committed:
enterprise: Silently fail config updates.

If a config item exists in enterprise config and we're not trying to
override, then setConfigItem returns.

kanishk98 committed:
settings: Show tooltip when setting disabled.

If setting has been locked by an admin, then a tooltip is shown in the
settings page.

kanishk98 commented:
Converted the branch to TS. Please have a look and tell me if this works well.

akashnimare commented:
@vsvipul, @priyank-p can you guys review?

vsvipul commented on app/resources/messages.ts:
```diff
@@ -22,6 +22,20 @@ class Messages {
 		\nYou can click here if you'd like to proceed with the connection.`;
 	}
+
+	enterpriseOrgError(length: number): any {
```
@kanishk98 You can create an interface here instead of any as you already know that it has title and content only. 
```
interface ErrorType {
    title: string;
    content: string;
}
```
and then use it here, instead of any. Same for orgRemovalError.

vsvipul commented on docs/Enterprise.md:
```diff
@@ -0,0 +1,26 @@
+and inside `C:\Program Files\Zulip-Desktop-Config` on Windows.
+It must be named `enterprise_config.json` in both cases. 
+
+To specify the preset organization you want to add for other users, you will need to
```
Change to "organizations".

priyank-p replied:
Considering the flow of the docs and the example below this should be left as `organization` because we give a example of specifying a single organization.

vsvipul commented on app/renderer/js/utils/enterprise-util.ts:
```diff
@@ -0,0 +1,69 @@
+let instance: null | EnterpriseUtil = null;
+
+class EnterpriseUtil {
+	enterpriseSettings: any;
```
Once the settings we'll allow are finalized, we'll need to remove this any.

kanishk98 replied:
Yeah, I'll leave a todo comment here for future work.

vsvipul commented on app/renderer/js/main.ts:
```diff
@@ -93,6 +95,7 @@ class ServerManagerView {
 	tabs: ServerOrFunctionalTab[];
 	functionalTabs: AnyObject;
 	tabIndex: number;
+	presetOrgs: any[];
```
We can create an interface for this too, if we know what it is going to contain.

akashnimare commented:
@priyank-p, @vsvipul, @kanishk98 let's take care of this PR and merge it asap so that we can include this in the next stable release.

kanishk98 committed:
typescript: Define DialogBoxError in messages.ts.

kanishk98 commented:
@akashnimare made the changes as suggested.

akashnimare commented:
> Document the new files made and the process of editing custom settings

@kanishk98 is this done?

kanishk98 commented:
> @kanishk98 is this done?

Yeah, it's at `docs/Enterprise.md`, isn't it?

akashnimare commented:
@kanishk98 should we show the name of the org here?

![image](https://user-images.githubusercontent.com/2263909/61901104-a5702700-af3c-11e9-839a-fde9e726ceb1.png)

Like -
```
Could not add the following organizations:
* Name of org
Please contact your system administrator
```

akashnimare commented:
The PR works as expected on macOS. @priyank-p, @vsvipul, @abhigyank can you folks also give it a try so that we can merge it?

ssvenn commented:
I think it would be better to look for global_config.json in the directory where the Zulip app is installed instead of hardcoding a path. It would probably be more flexible and it should survive updates as long as the file is only included in the archive as a template like ./examples/global_config.json

priyank-p commented:
@ssvenn so the reason we decided to choose the path we currently have is because it is easier for administrator to add the config in one place instead of having to do so for every single user. But I suppose it wouldn't be hard to implement support for placing the file in multiple location; @kanishk98 can you implement this it should be relatively easy to implement.

I think it better to look for it in user data directory instead of app directory as @ssvenn suggest because that what other apps do.

akashnimare commented:
> suppose it wouldn't be hard to implement support for placing the file in multiple locations; 

This might complicate things, IMO should keep only one location. We wanted to keep this file separate with app's file so chose that location. Let's keep it as it is for now and we can always change it later based on more feedbacks etc.

kanishk98 commented:
I agree with @akashnimare here, but I'd love to discuss the multiple locations thing for the next iteration of this feature. 

Are we allowing the user to specify the location for the files themselves or do we add another location on our own in the user directory and give preference to the admin's settings whenever there's a conflict?

ssvenn commented:
I figured looking for the configuration file in the program directory was just a more common and tidier implementation with one less directory in C:\Program Files\, the ability to work on computers where C:\ isn't the system drive, multiple instances and so on. It's a pretty minor thing and maybe only relevant for Windows. If you end up reading the config from many locations the order of precedence should probably be system directory, program directory, user directory.

akashnimare commented on docs/Enterprise.md:
```diff
@@ -0,0 +1,26 @@
+It must be named `enterprise_config.json` in both cases. 
+
+To specify the preset organization you want to add for other users, you will need to
+add the `json` shown below to the `enterprise_config.json`, replace `https://chat.zulip.org` to the
```
You need to update the name of this file to `global_config.json`.

akashnimare commented on docs/Enterprise.md:
```diff
@@ -0,0 +1,26 @@
+all users of your system, you can do so by creating an enterprise config file.
+The file should be placed at `/etc/zulip-desktop-config` for Linux and macOS computers
+and inside `C:\Program Files\Zulip-Desktop-Config` on Windows.
+It must be named `enterprise_config.json` in both cases. 
```
^^ `global_config.json`.

akashnimare commented on app/renderer/js/main.ts:
```diff
@@ -232,6 +240,59 @@ class ServerManagerView {
+		}
+	}
+
+	async initPresetOrgs(): Promise<void> {
```
Can you document this function? Add a comment or two.

akashnimare commented on app/renderer/js/main.ts:
```diff
@@ -232,6 +240,59 @@ class ServerManagerView {
 		this.toggleSidebar(showSidebar);
 	}
 
+	async queueDomain(domain: any): Promise<boolean> {
```
Please document this.

akashnimare commented on app/renderer/js/utils/enterprise-util.ts:
```diff
@@ -0,0 +1,70 @@
+import Logger = require('./logger-util');
+
+const logger = new Logger({
+	file: 'enterpise-util.log',
```
Typo - `enterpise`?

akashnimare commented on app/resources/messages.ts:
```diff
@@ -22,6 +27,20 @@ class Messages {
+
+	enterpriseOrgError(length: number): DialogBoxError {
+		return {
+			title: `Could not add ${length} ${length === 1 ? `organization` : `organizations`}`,
```
@kanishk98 can you update this as we discussed?

kanishk98 replied:
This is what it looks like now:

![image](https://user-images.githubusercontent.com/24617297/62161493-f5336180-b333-11e9-90c7-e091be096346.png)

Just checking a few things to make sure there are no problems with this. Will push in a bit.

kanishk98 replied:
I'll remove the length, that's unnecessary.

kanishk98 committed:
enterprise: Restrict EnterpriseUtil method calls.

* Add a boolean flag to EnterpriseUtil so unnecessary function calls
don't take place for non-admin users.

akashnimare commented:
As of now, the error message looks like this -

![image](https://user-images.githubusercontent.com/2263909/62166047-6c212800-b33d-11e9-9a6e-43f089502a5d.png)

We also need to add the org url here.

kanishk98 commented:
Works fine for me: 

![image](https://user-images.githubusercontent.com/24617297/62166270-f6698c00-b33d-11e9-914a-eeb470e60a59.png)

Try running `npm run clean-ts-files` first? Not doing so often doesn't reflect changes quickly for me.

vsvipul reviewed (approved):
Works fine for me. The app fails to even load if a wrong format is used for the config file. Leaving that, everything works fine.

akashnimare commented:
@kanishk98 great work 👌  This was the most demanded feature, glad we finished it. Thanks, everyone for the testing, inputs and reviews. @rishig would be great if you can look at the docs when you get time. @timabbott FYI.
